### PR TITLE
Enable prism by default (where supported)

### PR DESCRIPTION
**Pull request by damccorm**
Changes the default Python runner to be prism (when prism is supported). Excludes it from interactive environments for now.

There are a bunch of use cases automatically excluded, but this also required updating a bunch of tests. The 2 main breaking changes this introduces are:

1) Instead of raising Python-specific exceptions, when Prism jobs fail a `RuntimeError` is always raised. To fix any tests relying on the old error, change your exception catching logic to look for broader errors and use regexes to verify the contents of the error. There are lots of examples of this in this PR.
2) Prism names its metrics differently than the local runner. For this PR in Beam, I just kept all of those on the old FnApiRunner for now, but you could also just change the metric name you're looking for (this will be the eventual fix for these tests in this repo).

If this change breaks you in unexpected ways, please:

1) Pin to the `FnApiRunner` instead of the `DirectRunner`. There are lots of examples of this in this PR.
2) Comment in https://github.com/apache/beam/issues/34549 describing your broken use case.

Part of https://github.com/apache/beam/issues/34549

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 1]
```diff
@@ -405,7 +405,7 @@ def get_all_options(
     # sub-classes in the main session might be repeated. Pick last unique
     # instance of each subclass to avoid conflicts.
     subset = {}
-    parser = _BeamArgumentParser()
+    parser = _BeamArgumentParser(allow_abbrev=False)
```
Comment (damccorm) [inline sdks/python/apache_beam/options/pipeline_options.py:408]: This was causing all sorts of options because --output was getting assigned to --output_executable_path (https://github.com/apache/beam/blob/f6d209ff6a3ea638f11dfb96e52f7c21292dfbc1/sdks/python/apache_beam/options/pipeline_options.py#L1693C12-L1693C34) which meant the portable runner couldn't find prism

File: sdks/python/apache_beam/examples/complete/game/leader_board_test.py [hunk 2]
```diff
@@ -65,7 +65,7 @@ def test_leader_board_users(self):
       result = (
           self.create_data(p)
           | leader_board.CalculateUserScores(allowed_lateness=120))
-      assert_that(result, equal_to([]))
+      assert_that(result, equal_to([('user1_team1', 50), ('user2_team2', 2), ('user3_team3', 8), ('user4_team3', 5)]))
```
Comment (damccorm) [inline sdks/python/apache_beam/examples/complete/game/leader_board_test.py:68]: This should not be empty - Prism is handling this correctly, the FnApiRunner was dropping data

File: sdks/python/apache_beam/yaml/yaml_mapping_test.py [hunk 3]
```diff
@@ -175,7 +175,7 @@ def test_validate(self):
           label='Errors')
 
   def test_validate_explicit_types(self):
-    with self.assertRaisesRegex(TypeError, r'.*violates schema.*'):
+    with self.assertRaisesRegex(RuntimeError, r'TypeError.*violates schema.*'):
```
Comment (robertwb) [inline sdks/python/apache_beam/yaml/yaml_mapping_test.py:178]: How about asserting just Exception rather than that it must now be a RuntimeError? (For all of these.)

Comment (damccorm) [reply sdks/python/apache_beam/yaml/yaml_mapping_test.py:178]: Sure, I can get behind that

File: sdks/python/apache_beam/runners/portability/prism_runner.py [hunk 4]
```diff
@@ -331,9 +332,12 @@ def subprocess_cmd_and_endpoint(
     return (subprocess_cmd, f"localhost:{job_port}")
 
   def prism_arguments(self, job_port) -> typing.List[typing.Any]:
+    # TODO: Remove or expose log_level more nicely.
```
Comment (damccorm) [inline sdks/python/apache_beam/runners/portability/prism_runner.py]: TODO before merging

**Comment by damccorm** [comment]
Most remaining flakes should be fixed by https://github.com/apache/beam/pull/35337 with the next release, pausing work until then

**Comment by damccorm** [comment]
Current persistent (or possibly persistent) test failures:

**[beam_PreCommit_Python_ML](https://github.com/apache/beam/actions/runs/15831111506/job/44623727892?pr=34612#logs)** and **[Python Unit Tests](https://github.com/apache/beam/actions/runs/15831111847/job/44623728997?pr=34612#logs)**:

- `apache_beam.ml.ts.util_test.PeriodicStreamTest` -  a few failures, but this is a bad test. See https://github.com/apache/beam/pull/35300#issuecomment-2997517699 - if needed can fix here, but I'd rather we make some changes upstream

**[beam_PreCommit_Python](https://github.com/apache/beam/actions/runs/15831111491/job/44623727865?pr=34612#logs)**:

- apache_beam/io/textio_test.py::TextSinkTest::test_write_pipeline_footer - not sure what the issue is yet

**[beam_PreCommit_Python_Transforms](https://github.com/apache/beam/actions/runs/15831111519/job/44623727853?pr=34612#logs)**:

- apache_beam/transforms/userstate_test.py::StatefulDoFnOnDirectRunnerTest::test_stateful_set_state_clean_portably 

**Linting/formatting** - I'll deal with this later

There are also some other flakes, which may or may not be related to this change

**Comment by damccorm** [comment]
R: @shunping

File: sdks/python/apache_beam/io/fileio_test.py [hunk 5]
```diff
@@ -106,7 +106,7 @@ def test_match_files_one_directory_failure1(self):
     files.append(self._create_temp_file(dir=directories[0]))
     files.append(self._create_temp_file(dir=directories[0]))
 
-    with self.assertRaises(beam.io.filesystem.BeamIOError):
+    with self.assertRaises(Exception):
```
Comment (shunping) [inline sdks/python/apache_beam/io/fileio_test.py]: When a Runtime Exception is raised in prism, maybe we can improve this to use [assertRaisesRegex](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex) instead to match exception message? Just like some example in `sdks/python/apache_beam/ml/inference/pytorch_inference_test.py` below.

File: sdks/python/apache_beam/ml/inference/tensorflow_inference_test.py [hunk 6]
```diff
@@ -128,7 +128,8 @@ def test_predict_tensor_with_batch_size(self):
     model = _create_mult2_model()
     model_path = os.path.join(self.tmpdir, f'mult2_{uuid.uuid4()}.keras')
     tf.keras.models.save_model(model, model_path)
-    with TestPipeline() as pipeline:
+    # FnApiRunner guarantees large batches, which this pipeline assumes
```
Comment (shunping) [inline sdks/python/apache_beam/ml/inference/tensorflow_inference_test.py:131]: This could be due to the fact that prism is using accelerated processing time by default. The processing-time timer in BatchElements may not work as expected.

PR #35202 added an experimental flag for using real-time processing time, which may help to resolve this problem.

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 7]
```diff
@@ -112,9 +118,25 @@ def visit_transform(self, applied_ptransform):
...
+        all_options = options.get_all_options()
         self.supported_by_prism_runner = True
-        pipeline.visit(self)
+        # TODO(https://github.com/apache/beam/issues/33623): Prism currently
+        # double fires on AfterCount trigger, once appropriately, and once
+        # incorrectly at the end of the window. This if condition could be
+        # more targeted, but for now we'll just ignore all unsafe triggers.
+        if pipeline.allow_unsafe_triggers:
+          self.supported_by_prism_runner = False
+        # TODO(https://github.com/apache/beam/issues/33623): Prism currently
```
Comment (shunping) [inline sdks/python/apache_beam/runners/direct/direct_runner.py:130]: What else problems do you face here when using prism in interactive mode?

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 8]
```diff
@@ -136,41 +176,40 @@ def visit_transform(self, applied_ptransform):
...
+            state_specs, timer_specs = userstate.get_dofn_specs(dofn)
             for timer in timer_specs:
               if timer.time_domain == TimeDomain.REAL_TIME:
                 self.supported_by_prism_runner = False
 
-    tryingPrism = False
+            for state in state_specs:
+              if isinstance(state, userstate.CombiningValueStateSpec):
+                self.supported_by_prism_runner = False
+        # TODO(https://github.com/apache/beam/issues/33623): Prism seems to
```
Comment (shunping) [inline sdks/python/apache_beam/runners/direct/direct_runner.py:187]: Interesting, I recently fixed a bug when prism handles sliding window (#35385). Just want to reference here in case it is related.

File: sdks/python/apache_beam/transforms/util_test.py [hunk 9]
```diff
@@ -254,8 +254,8 @@ def test_constant_batch_no_metrics(self):
     self.assertEqual(len(results["distributions"]), 0)
 
   def test_grows_to_max_batch(self):
-    # Assumes a single bundle...
-    with TestPipeline() as p:
+    # Assumes a single bundle, so we pin to the FnApiRunner
```
Comment (shunping) [inline sdks/python/apache_beam/transforms/util_test.py:257]: Do we have a reliable way of getting elements in a single bundle? Maybe only for testing purpose. This seems to be a recurrent issue I meet when I am working on some other tests. 

Alternatively, we may be able to change the tests so that they won't rely on this assumption, but it is not working all the time.

**Review by shunping — APPROVED** [review]
Thank you for making this big step for prism! I really appreciate the thorough and detailed work you've put into this. 

The overall PR looks solid. I've left a few minor comments for your consideration.

**Comment by damccorm** [comment]
There were failing postcommit tests, but not on prism, so I'm not going to block merging

**Comment by Abacn** [comment]
There are tests failing after this PR, could it be related?

https://github.com/apache/beam/actions/workflows/beam_PostCommit_Python_Xlang_Gcp_Direct.yml?query=event%3Aschedule

https://github.com/apache/beam/actions/workflows/beam_PostCommit_Yaml_Xlang_Direct.yml?query=event%3Aschedule

**Comment by damccorm** [comment]
> There are tests failing after this PR, could it be related?
> 
> https://github.com/apache/beam/actions/workflows/beam_PostCommit_Python_Xlang_Gcp_Direct.yml?query=event%3Aschedule
> 
> https://github.com/apache/beam/actions/workflows/beam_PostCommit_Yaml_Xlang_Direct.yml?query=event%3Aschedule

I think it likely is - https://github.com/apache/beam/pull/35608 tested these workflows with extended timeouts and they still failed. Temporarily reverting the core of this change here - https://github.com/apache/beam/pull/35610
